go-sdk/internal/utxo/encryption.go

```diff
@@ -0,0 +1,169 @@
+// as described: https://drive.google.com/file/d/1EVrP3DzoGbmzkRmYnyEDcIQcXVU7GlOd/view
+// The encryption and decryption functions are compatible with the circom implementations,
+// meaning the cipher texts encrypted by the circuit in circuits/lib/encrypt.circom can
+// be decrypted by the PoseidonDecrypt function. And vice versa.
+func PoseidonEncrypt(msg []*big.Int, key []*big.Int, nonce *big.Int) ([]*big.Int, error) {
+	if len(key) != 2 {
+		return nil, fmt.Errorf("the key must have 2 elements, but got %d", len(key))
+	}
+
+	// the size of the message array must be a multiple of 3
+	// pad with 0 if necessary
+	length := len(msg)
+	if length%3 > 0 {
+		length += 3 - (length % 3)
+	}
+	message := make([]*big.Int, length)
+	copy(message, msg)
+	for idx := len(msg); idx < length; idx++ {
+		message[idx] = big.NewInt(0)
+	}
+
+	// Create the initial state
+	// S = (0, kS[0], kS[1], N + l * 2^128)
+	l := big.NewInt(int64(len(msg)))
+	state := []*big.Int{big.NewInt(0), key[0], key[1], big.NewInt(0).Add(nonce, big.NewInt(0).Mul(l, &two128))}
+
+	cipherText := make([]*big.Int, length+1)
+	var err error
+
+	n := length / 3
+	i := 0
+	for ; i < n; i++ {
+		// Iterate Poseidon on the state
+		state, err = poseidon.HashEx(state, 4)
+		if err != nil {
+			return nil, err
+		}
+
+		// Modify the state for the next round
+		// state[1] = addMod(message[i * 3], state[1]);
+		state[1] = ff.NewElement().Add(ff.NewElement().SetBigInt(message[i*3]), ff.NewElement().SetBigInt(state[1])).ToBigIntRegular(state[1])
+		// state[2] = addMod(message[i * 3 + 1], state[2]);
+		state[2] = ff.NewElement().Add(ff.NewElement().SetBigInt(message[i*3+1]), ff.NewElement().SetBigInt(state[2])).ToBigIntRegular(state[2])
+		// state[3] = addMod(message[i * 3 + 2], state[3]);
+		state[3] = ff.NewElement().Add(ff.NewElement().SetBigInt(message[i*3+2]), ff.NewElement().SetBigInt(state[3])).ToBigIntRegular(state[3])
+
+		// Record the three elements of the encrypted message
+		cipherText[i*3] = state[1]
+		cipherText[i*3+1] = state[2]
+		cipherText[i*3+2] = state[3]
+	}
+
+	// Iterate Poseidon on the state one last time
+	state, err = poseidon.HashEx(state, 4)
+	if err != nil {
+		return nil, err
+	}
+
+	// Record the last ciphertext element
+	cipherText[i*3] = state[1]
+
+	return cipherText, nil
+}
+
+// the "cipherText" parameter must be the output of the PoseidonEncrypt function, with potentially
+// extra elements due to the padding of the plain text messages with 0s to make the length to be 3n.
+//
+// the "length" parameter must be the length of the original message
+func PoseidonDecrypt(cipherText []*big.Int, key []*big.Int, nonce *big.Int, length int) ([]*big.Int, error) {
+	if len(key) != 2 {
+		return nil, fmt.Errorf("the key must have 2 elements, but got %d", len(key))
+	}
+	// length of the cipher text must be 3n+1
+	if len(cipherText)%3 != 1 {
+		return nil, fmt.Errorf("the length of the cipher text must be 3n+1, but got %d", len(cipherText))
```

Comment:
the extra padding including the last element in the cipher text are all for safety checks, to ensure that unless the right keys are used, the decrypt will fail. otherwise you'll always get some number even if the wrong keys are used.